src/bluesky/simulators.py

```diff
@@ -104,3 +106,195 @@ async def check_limits_async(plan):
             else:
                 warn(f"{obj.name} has no check_value() method" f" to check if {msg.args[0]} is within its limits.")  # noqa: B028
                 ignore.append(obj)
+
+
+class RunEngineSimulator:
+    """This class facilitates testing of a Bluesky plan by recording bluesky messages and
+     injecting responses according to the bluesky Message Protocol (see bluesky docs for details).
+
+    Basic usage consists of
+    1) Registering various handlers to respond to anticipated messages in the experiment plan and fire any
+    needed callbacks.
+    2) Calling simulate_plan()
+    3) Examining the returned message list and making asserts against them
+
+    An example usage:
+    >>> def my_plan():
+    ...     pass
+
+    >>> sim = RunEngineSimulator()
+    >>> messages = sim.simulate_plan(my_plan())
+    >>>
+
+
+    """
+
+    GROUP_ANY = "any"
+
+    def __init__(self):
+        self.message_handlers = []
+        self.callbacks = {}
+        self.next_callback_token = 0
+        self.return_value: Any
+
+    def add_handler_for_callback_subscribes(self):
+        """Add a handler that registers all the callbacks from subscribe messages so we can call them later.
+        You will need to call this as one of the first things if you wish to fire callbacks from the simulator.
+        """
+        self.message_handlers.append(
+            _MessageHandler(
+                lambda msg: msg.command == "subscribe",
+                lambda msg: self._add_callback(msg.args),
+            )
+        )
+
+    def add_handler(
+        self,
+        handler: Callable[[Msg], object],
+        commands: Union[str, Sequence[str]],
+        obj_name: Optional[str] = None,
```

Comment:
maybe make `obj_name` and `predicate` key-word only?  Would it make sense to merge them into one kwarg called "filter" which which is either a string or a `Callable[[Msg], bool]`.  It is not obvious to me that `or` is unambigiously the correct way to combine the two filters so my merging into one input we avoid having to make that choice.